Implement unknown6 and signature encryption

Gabrola opened the pull request:
I've implemented the signature encryption algorithm in c# so there is no need to p/invoke the C encrypt.dll
I've confirmed that this works on my bot and everything works perfectly.

Gabrola committed:
Implement unknown6 and signature encryption

I've implemented the signature encryption algorithm in c# so there is no
need to p/invoke the C encrypt.dll

wertzui commented on PokemonGo.RocketAPI/POGOProtos/Networking/Envelopes/Signature.cs:
```diff
@@ -0,0 +1,2580 @@
+    static SignatureReflection() {
+      byte[] descriptorData = global::System.Convert.FromBase64String(
+          string.Concat(
+            "Ci9QT0dPUHJvdG9zL05ldHdvcmtpbmcvRW52ZWxvcGVzL1NpZ25hdHVyZS5w",
```
Why are you pasting the string as base64 here?

Gabrola replied:
This whole file is compiled, I didn't paste anything in there. If you're curious, this is a protofile descriptor which is needed when you are working with message types not known at compile-time or you want to access fields at runtime using reflection. Either way, this is not needed in this application since we are already using the compiled classes, it's only there because the protobuf compiler did.

Gabrola committed:
Fix empty GetMapObjects

Gabrola committed:
Typo

Panthere commented:
Wow, thats quite an amazing port from  the C++ "_"

Good job.
I think I like this one a bit less than the other, but if the crypt.cs was included in the other I think it would work quite good. Perhaps a class for building the u6 sig?

ArcticXWolf commented on PokemonGo.RocketAPI/Helpers/Utils.cs:
```diff
@@ -9,5 +12,37 @@ public static ulong FloatAsUlong(double value)
+            var locationBytes = new List<byte>();
+            locationBytes.AddRange(BitConverter.GetBytes(lat).Reverse());
+            locationBytes.AddRange(BitConverter.GetBytes(lng).Reverse());
+            locationBytes.AddRange(BitConverter.GetBytes(alt).Reverse());
```
Why is the reversing of the byte array necessary? :)

Gabrola replied:
Values are stored in little endian format in memory on most computer architectures, including x86, which is reverse order from a human perspective. For some reason Niantic decided to use big-endian order for their hashing function so we have to comply and reverse the normal order. So a value of 0x01020304 is stored in memory as 04 03 02 01 so we have to reverse it. In the python implementation they're converting the value to a human-readable hex string and then split that up into a byte array.

ArcticXWolf replied:
Ah, okay I though GetBytes automatically converts that back, but so this makes sense. Thank you :)

Gabrola committed:
Use AuthInfo as seed when AuthTicket is not available yet

https://github.com/keyphact/pgoapi/commit/7a0666fc534b4ec14741922214ac9795258685bf

Gabrola committed:
Apply unknown6 to inital request too
